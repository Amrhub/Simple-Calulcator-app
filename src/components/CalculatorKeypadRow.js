import React from 'react';
import PropTypes from 'prop-types';

const KeypadRow = (props) => {
  const { values } = props;
  const { callCalculate } = props;

  const clickHandler = (e) => {
    const ButtonName = e.target.innerText;
    callCalculate(ButtonName);
  };
  return (
    <div className="calculator__keypad-row">
      {values.map((value) => (
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          type="button"
          key={value}
          onClick={clickHandler}
        >
          {value}
        </button>
      ))}
    </div>
  );
};

KeypadRow.propTypes = {
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
  callCalculate: PropTypes.func.isRequired,
};

export default KeypadRow;
