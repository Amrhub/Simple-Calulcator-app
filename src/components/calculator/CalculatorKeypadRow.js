import React from 'react';
import PropTypes from 'prop-types';

// eslint-disable-next-line react/prefer-stateless-function
export default class KeypadRow extends React.Component {
  render() {
    const { values } = this.props;
    return (
      <div className="calculator__keypad-row">
        {values.map((value) => (
          <button
            className="calculator__keypad-key flex-center-center"
            type="button"
            key={value}
          >
            {value}
          </button>
        ))}
      </div>
    );
  }
}

// eslint-disable-next-line react/forbid-prop-types
KeypadRow.propTypes = { values: PropTypes.array.isRequired };
