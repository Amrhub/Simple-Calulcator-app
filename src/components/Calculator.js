import React from 'react';
import './calculator.scss';
import KeypadRow from './CalculatorKeypadRow';

export default class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = { result: 0 };
  }

  render() {
    const { result } = this.state;
    return (
      <div className="calculator flex-col-center">
        <div className="calculator__display">{result}</div>
        <div className="calculator__keypad">
          <KeypadRow values={['AC', '+/-', '%', 'รท']} />
          <KeypadRow values={['7', '8', '9', 'x']} />
          <KeypadRow values={['4', '5', '6', '-']} />
          <KeypadRow values={['1', '2', '3', '+']} />
          <KeypadRow values={['0', '.', '=']} />
        </div>
      </div>
    );
  }
}
