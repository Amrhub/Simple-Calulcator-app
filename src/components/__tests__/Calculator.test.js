import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import Calculator from '../Calculator';
import CalculatorKeypadRow from '../CalculatorKeypadRow';

const callCalculate = jest.fn(); // create a mock function

describe('Calculator', () => {
  it('Calculator component renders correctly', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('CalculatorKeypadRow renders correctly', () => {
    const tree = renderer
      .create(
        <CalculatorKeypadRow
          values={['AC', '+/-', '%', 'รท']}
          callCalculate={callCalculate}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Calculator user interaction', () => {
  it('when user hits a "7" it renders "7"', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText(/7/i));

    const display = screen.getByTestId('display');

    expect(display).toHaveTextContent(/7/i);
  });

  it('when user hits a "4" then hits "5" it renders "45"', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText(/4/i));
    userEvent.click(screen.getByText(/5/i));

    const display = screen.getByTestId('display');

    expect(display).toHaveTextContent(/45/i);
  });

  it('when user hits a "4" then hits "+" then hits "5" then hits "=" it renders "9"', () => {
    render(<Calculator />);

    userEvent.click(screen.getByText(/4/i));
    userEvent.click(screen.getByText('+'));
    userEvent.click(screen.getByText(/5/i));
    userEvent.click(screen.getByText('='));

    const display = screen.getByTestId('display');

    expect(display).toHaveTextContent(/9/i);
  });
});
