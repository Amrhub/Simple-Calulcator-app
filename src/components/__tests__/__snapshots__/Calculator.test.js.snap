// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator Calculator component renders correctly 1`] = `
<main
  className="calculator-container "
>
  <h2
    className="calculator-container__heading"
    data-testid="calculator-heading"
  >
    Let's do some math!
  </h2>
  <div
    className="calculator flex-col-center"
  >
    <div
      className="calculator__display number-font-family"
      data-testid="display"
    >
      0
    </div>
    <div
      className="calculator__keypad"
    >
      <div
        className="calculator__keypad-row"
      >
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          AC
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          +/-
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          %
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          รท
        </button>
      </div>
      <div
        className="calculator__keypad-row"
      >
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          7
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          8
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          9
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          x
        </button>
      </div>
      <div
        className="calculator__keypad-row"
      >
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          5
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          6
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          -
        </button>
      </div>
      <div
        className="calculator__keypad-row"
      >
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          +
        </button>
      </div>
      <div
        className="calculator__keypad-row"
      >
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          0
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          .
        </button>
        <button
          className="calculator__keypad-key flex-center-center number-font-family"
          onClick={[Function]}
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>
</main>
`;

exports[`Calculator CalculatorKeypadRow renders correctly 1`] = `
<div
  className="calculator__keypad-row"
>
  <button
    className="calculator__keypad-key flex-center-center number-font-family"
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
  <button
    className="calculator__keypad-key flex-center-center number-font-family"
    onClick={[Function]}
    type="button"
  >
    +/-
  </button>
  <button
    className="calculator__keypad-key flex-center-center number-font-family"
    onClick={[Function]}
    type="button"
  >
    %
  </button>
  <button
    className="calculator__keypad-key flex-center-center number-font-family"
    onClick={[Function]}
    type="button"
  >
    รท
  </button>
</div>
`;
